src/persist-types/src/arrow.rs

```diff
@@ -0,0 +1,581 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A [protobuf] representation of [Apache Arrow] arrays.
+//!
+//! # Motivation
+//!
+//! Persist can store a small amount of data inline at the consensus layer.
+//! Because we are space constrained, we take particular care to store only the
+//! data that is necessary. Other Arrow serialization formats, e.g. [Parquet]
+//! or [Arrow IPC], include data that we don't need and would be wasteful to
+//! store.
+//!
+//! # Design Considerations
+//!
+//! Initially this was implemented as a "protobuf port" of
+//! [`arrow::array::ArrayData`], with the goal of zero-copy deserialization
+//! via [`arrow::buffer::Buffer`]s and [`bytes::Bytes`]. This doesn't work
+//! though because Arrow needs the underlying Buffers to be aligned for the
+//! type of data stored, e.g. an array of `u64`s needs to be 8 byte aligned.
+//! While it is technically possible to make this work, it is much easier to
+//! structure the data in protobuf and use native types.
```

Comment:
That does work! But between the possibility of copying the buffer and some space overhead I opted for the current approach. I saved the `ArrayData` impl though, more than happy to revisit it if you like that approach better

Reply:
Well, I suspect that one advantage of the `ArrayData` / `build_aligned` approach is that it would be far fewer lines of code. But you are the one who knows for sure - I've never tried it, but you have the branch!

I can believe that a better encoding and saving some copies is worth whatever extra code we're paying for it... but if so I might adjust the wording here, since it sounds like the main blocker was that it was hard to do and not that it was expensive!

Reply:
Here's the `ArrayData` approach ([gist](https://gist.github.com/ParkMyCar/2e26d3218d4d91c3276bfae94cebf7da))! It's significantly less code but ends up on average having a ~40% larger encoded size than the current approach. Which to me seems like juuuust on the line of acceptable and I'm not sure which way to go

Reply:
Cool -- thank you for putting that together! I think I feel good about that tradeoff, but I'm happy to defer to @danhhz who has the best instincts about inline writes still I think!

Reply:
We talked offline. IMO either approach is shippable. I lean slightly toward ArrayData given that:
- It's less code (and it already feels a bit bad to be inventing a new arrow serialization format)
- It's probably easier to get back 40% increased encoding size (gzip?) vs cpu work for varint decoding. (Otoh, these are small so probably both are fine.)